### PR TITLE
Add dependencies to trino-server

### PR DESCRIPTION
**Pull request by nineinchnick**
This PR adds dependencies to trino-server generated using the following command:
```shell
./mvnw provisio:generateDependencies \
  -DskipTests \
  -DpomFile=generated.xml \
  -pl core/trino-server
```
and then manually comparing `pom.xml` and `generated.xml` to copy dependencies but retain variables, comments, and formatting.

Such dependencies can be useful for other Maven plugins, like `license-maven-plugin`. They would also [show up in Maven Central ](https://mvnrepository.com/artifact/io.trino/trino-server/356) and would give a better idea of what components are contained in the tarball.

Since the dependencies still have to maintained semi-manually, to keep them up to date, I suggest adding another job in the pipeline that would run the `provisio:validateDependencies` mojo and catch when any changes in dependencies. On the other hand, it only needs to be run once before a release, so maybe it should not block the pipeline.

It does print an actionable message:
```
Error:  Failed to execute goal ca.vanzyl.provisio.maven.plugins:provisio-maven-plugin:1.0.16:validateDependencies (default-cli) on project trino-server: Missing dependencies: io.airlift:launcher:tar.gz:properties:205, io.airlift:launcher:tar.gz:bin:205 -> [Help 1]
```

Where would be a good place to add some description of this workflow? In `DEVELOPMENT.md`?

### CODE REVIEW

**Comment by hashhar** [comment]
Can the plugins/tools which require this information be run in a dirty-tree with the generated pom.xml instead of comitting the generated pom? IIUC the generated pom isn't going to be shipped with Trino - it's only required for auxiliary processes?

i.e. use separate representations for code vs humans? Also how does the generated pom handle multiple versions of same dependency (possible because we have classloader isolation per plugin in Trino)? If it picks one version then we are losing information about the other version.

e.g. we could use two different versions of Jackson and one of them may have a CVE while the one which gets written to generated pom might be the later one and it'd look as if we have no CVEs?

**Comment by nineinchnick** [comment]
Excellent questions!

> Can the plugins/tools which require this information be run in a dirty-tree with the generated pom.xml instead of comitting the generated pom? IIUC the generated pom isn't going to be shipped with Trino - it's only required for auxiliary processes?

It could, but only for Trino. If we'd want to use it in any other project that repackages the Trino server tarball using Provisio, then the pom associated with the tarball needs to include all dependencies.

But I guess the pom can be replaced in a local clone and then installed into the local Maven repository. This complicates the build process a bit but I'm ok with both approaches. @electrum @martint @dain @findepi WDYT?

> 
> i.e. use separate representations for code vs humans? Also how does the generated pom handle multiple versions of same dependency (possible because we have classloader isolation per plugin in Trino)? If it picks one version then we are losing information about the other version.
> 
> e.g. we could use two different versions of Jackson and one of them may have a CVE while the one which gets written to generated pom might be the later one and it'd look as if we have no CVEs?

Both new Mojos in Provisio detect multiple versions of the same direct dependency and throw an error. We should be able to update direct dependencies to the same version. If this happens in transitive dependencies, any tool/plugin processing them (eg. license-maven-plugin) should be able to deal with it.

File: core/trino-server/pom.xml [hunk 1]
```diff
@@ -26,6 +26,2440 @@
         <process-name>${project.artifactId}</process-name>
     </properties>
 
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>io.trino</groupId>
+                <artifactId>trino-array</artifactId>
+                <version>357-SNAPSHOT</version>
```
Comment (findepi) [inline core/trino-server/pom.xml:34]: ins't the same written in top level pom?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -40,6 +40,10 @@ jobs:
         run: |
           export MAVEN_OPTS="${MAVEN_INSTALL_OPTS}"
           $RETRY ./mvnw install -B -V -T C1 -DskipTests -P ci -pl '!:trino-server-rpm'
+      - name: Test Server Dependencies
+        run: |
+          export MAVEN_OPTS="${MAVEN_INSTALL_OPTS}"
+          ./mvnw provisio:validateDependencies ${MAVEN_FAST_INSTALL} -pl :trino-server
```
Comment (findepi) [inline .github/workflows/ci.yml:46]: Add a code level comment what it does and why this is important.

(i.e. what to do when this fails and what should i consider if i wanted to change/remove these lines?)

File: core/trino-server/pom.xml [hunk 3]
```diff
@@ -26,6 +26,2440 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>aopalliance</groupId>
+            <artifactId>aopalliance</artifactId>
+            <version>1.0</version>
+        </dependency>
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-core</artifactId>
+            <version>1.2.3</version>
```
Comment (findepi) [inline core/trino-server/pom.xml:1585]: It's a transitive dependency.
Is your goal to have something like yarn lock?

**Comment by findepi** [comment]
> Since the dependencies still have to maintained semi-manually, to keep them up to date, I suggest adding another job in the pipeline that would run the provisio:validateDependencies mojo and catch when any changes in dependencies. On the other hand, it only needs to be run once before a release, so maybe it should not block the pipeline.

if they are to be kept up to date, they need to be updated when the change happens.
Release process should be as lightweight as possible.

> It could, but only for Trino. If we'd want to use it in any other project that repackages the Trino server tarball using Provisio, then the pom associated with the tarball needs to include all dependencies.
> 
> But I guess the pom can be replaced in a local clone and then installed into the local Maven repository.

Sounds reasonable to me.

> > e.g. we could use two different versions of Jackson and one of them may have a CVE while the one which gets written to generated pom might be the later one and it'd look as if we have no CVEs?
> 
> Both new Mojos in Provisio detect multiple versions of the same direct dependency and throw an error. We should be able to update direct dependencies to the same version. 

I don't think it's a safe assumption. Some libraries are/were not backwards compatible. For example, Guava is _now_ stated to be backwards compatible but some plugins were stuck on some pre-compatibility version like 19. Furthermore, currentl some plugins purposefully pull different versions of the same dependency. Compare Phoenix (4) vs Phoenix 5 plugins.

The plugin subsystem is important IMO primarily because it allows you to manage large number of dependencies without classpath hell (even more than isolation it provides), and we shouldn't assume the classpath hell is 'resolveble'. 

cc @sopel39 @kokosing

**Comment by nineinchnick** [comment]
> I don't think it's a safe assumption. Some libraries are/were not backwards compatible. For example, Guava is _now_ stated to be backwards compatible but some plugins were stuck on some pre-compatibility version like 19. Furthermore, currentl some plugins purposefully pull different versions of the same dependency. Compare Phoenix (4) vs Phoenix 5 plugins.
> 
> The plugin subsystem is important IMO primarily because it allows you to manage large number of dependencies without classpath hell (even more than isolation it provides), and we shouldn't assume the classpath hell is 'resolveble'.

Just to make sure, we're still talking only about direct dependencies? `plugin/trino-phoenix/pom.xml` has `org.apache.phoenix:phoenix-client:embedded:4.14.1-HBase-1.4` and `plugin/trino-phoenix5/pom.xml` has `org.apache.phoenix:phoenix-client-embedded-hbase-2.2:5.1.1`. Is it just a coincidence that the `artifactId` here is different?

File: core/trino-server/pom.xml (hunk 3 above)
Comment (nineinchnick) [reply core/trino-server/pom.xml:1585]: No, there should only be direct dependencies here. I think I messed this up in Provisio, let me check this and get back to you.

**Comment by kokosing** [comment]
I don't think it is the way to go. As others already mentioned, each plugin has a separate classloader so it can use whatever they like. Even if we would not have a conflict today we might have it tomorrow. I think we should process (analyze) each plugin separately. Is it possible to extract from provisio a list of plugins and then generate pom.xml files for them for your purposes. So you could have a list of pom.xml files instead of single one big sack?

**Comment by nineinchnick** [comment]
> I don't think it is the way to go. As others already mentioned, each plugin has a separate classloader so it can use whatever they like. Even if we would not have a conflict today we might have it tomorrow. I think we should process (analyze) each plugin separately. Is it possible to extract from provisio a list of plugins and then generate pom.xml files for them for your purposes. So you could have a list of pom.xml files instead of single one big sack?

The problem I'm trying to solve is when Provisio uses the Trino tarball as a base. The pom associated with that tarball doesn't contain any info on what's in the tarball itself. Provisio doesn't have any other index, I would have to inspect every jar in the tarball. I was trying to avoid building tools to process a tree of dependencies if Maven already can do this.

Looks like I included transitive dependencies right now and we only need direct dependencies. I'll close this PR until I can figure this out.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (nineinchnick) [reply .github/workflows/ci.yml:46]: Added a comment. The error message would list missing dependencies, like so:
```
Missing dependencies: io.airlift:launcher:tar.gz:properties:205
```
Is this actionable enough?

File: core/trino-server/pom.xml (hunk 1 above)
Comment (nineinchnick) [reply core/trino-server/pom.xml:34]: Removed.

File: core/trino-server/pom.xml (hunk 3 above)
Comment (nineinchnick) [reply core/trino-server/pom.xml:1585]: Removed all transitive dependencies.

**Comment by findepi** [comment]
>  I would have to inspect every jar in the tarball. 

That could actually be easier, since each such jar should contain META-INF/maven.properties (or sth like that) with coordinates.

**Pull request closed without merging**